File: authentication/tasks/backup.js
// backup.js

const { exec } = require('child_process');
const path = require('path');
const fs = require('fs');
const dotenv = require('dotenv');
const logger = require('../middlewares/logger'); // Ensure logger is properly configured

// Load environment variables from .env file
dotenv.config();

const DATABASE_NAME = 'PoGo_App_Users'; // Your actual database name
const COLLECTION_NAME = 'users'; // The collection you want to back up
const DATABASE_URL = process.env.DATABASE_URL || 'mongodb://localhost:27017';
const BACKUP_DIRECTORY = process.env.BACKUP_DIRECTORY || path.join(__dirname, '../backups'); // Ensure this directory exists

/**
 * Create a database backup.
 */
const createBackup = () => {
    // Ensure the backup directory exists
    if (!fs.existsSync(BACKUP_DIRECTORY)) {
        fs.mkdirSync(BACKUP_DIRECTORY, { recursive: true });
        logger.info(`Created backup directory: ${BACKUP_DIRECTORY}`);
    }

    const date = new Date();
    const timestamp = `${date.getFullYear()}_${(date.getMonth() + 1).toString().padStart(2, '0')}_${date.getDate().toString().padStart(2, '0')}`;
    const backupPath = path.join(BACKUP_DIRECTORY, `${DATABASE_NAME}_${timestamp}.gz`);

    // Log the backup path to verify
    logger.info(`Backup path: ${backupPath}`);

    // Construct the mongodump command to output a single archive file
    const command = `mongodump --uri="${DATABASE_URL}" --db=${DATABASE_NAME} --collection=${COLLECTION_NAME} --archive=${backupPath} --gzip`;

    // Log the command to verify
    logger.debug(`Executing command: ${command}`);

    // Execute the mongodump command
    exec(command, (error, stdout, stderr) => {
        if (error) {
            logger.error(`Backup failed: ${error.message}`);
        }
        if (stderr) {
            logger.error(`Backup stderr: ${stderr}`);
        }

        // Always log and verify the result, even if there was an error
        if (fs.existsSync(backupPath)) {
            logger.info(`Backup created successfully at: ${backupPath}`);
        } else {
            logger.error(`Backup file not found: ${backupPath}`);
        }

        // Ensure retention management is called no matter what
        logger.info('Starting retention management...');
        manageRetention();
    });
};

/**
 * Manage backup retention by following a Grandfather-Father-Son scheme.
 * - Daily (Son): Keep daily backups for the last 30 days
 * - Monthly (Father): Keep monthly backups (1st of each month) for 12 months
 * - Yearly (Grandfather): Keep yearly backups (1st Jan) for 5 years
 */
const manageRetention = () => {
    // Retention periods
    const dailyRetentionDays = 30;      // Keep daily backups for 30 days
    const monthlyRetentionMonths = 12;  // Keep monthly backups for 12 months
    const yearlyRetentionYears = 5;     // Keep yearly backups for 5 years

    const now = new Date();

    fs.readdir(BACKUP_DIRECTORY, (err, files) => {
        if (err) {
            logger.error(`Failed to read backup directory: ${err.message}`);
            return;
        }

        files.forEach((file) => {
            if (path.extname(file) !== '.gz') return; // Only consider .gz backups

            const match = file.match(/(\d{4})_(\d{2})_(\d{2})/);
            if (!match) {
                logger.error(`Filename does not match expected format: ${file}`);
                return;
            }

            const [_, yearStr, monthStr, dayStr] = match;
            const year = parseInt(yearStr, 10);
            const month = parseInt(monthStr, 10) - 1; // zero-based for JS months
            const day = parseInt(dayStr, 10);

            const fileDate = new Date(year, month, day);
            if (isNaN(fileDate)) {
                logger.error(`Invalid date parsed for file: ${file}`);
                return;
            }

            // Determine backup type based on date
            const isYearlyBackup = (month === 0 && day === 1);    // January 1st
            const isMonthlyBackup = (!isYearlyBackup && day === 1); 
            const isDailyBackup   = (!isYearlyBackup && !isMonthlyBackup);

            // Calculate ages
            const ageInDays = Math.floor((now - fileDate) / (1000 * 60 * 60 * 24));
            const ageInMonths = (now.getFullYear() - year) * 12 + (now.getMonth() - month);
            const ageInYears = now.getFullYear() - year;

            logger.debug(`Evaluating file: ${file}`);
            logger.debug(`Type: ${isYearlyBackup ? 'Yearly' : isMonthlyBackup ? 'Monthly' : 'Daily'}`);
            logger.debug(`Age: ${ageInDays} days, ${ageInMonths} months, ${ageInYears} years`);

            let shouldDelete = false;

            if (isDailyBackup && ageInDays > dailyRetentionDays) {
                logger.info(`Daily backup ${file} is older than ${dailyRetentionDays} days. Deleting.`);
                shouldDelete = true;
            } else if (isMonthlyBackup && ageInMonths > monthlyRetentionMonths) {
                logger.info(`Monthly backup ${file} is older than ${monthlyRetentionMonths} months. Deleting.`);
                shouldDelete = true;
            } else if (isYearlyBackup && ageInYears > yearlyRetentionYears) {
                logger.info(`Yearly backup ${file} is older than ${yearlyRetentionYears} years. Deleting.`);
                shouldDelete = true;
            } else {
                logger.info(`Backup ${file} does not meet deletion criteria.`);
            }

            if (shouldDelete) {
                deleteFile(path.join(BACKUP_DIRECTORY, file));
            }
        });
    });
};

/**
 * Delete a file and log the result.
 * @param {string} filePath - The path to the file to delete.
 */
const deleteFile = (filePath) => {
    fs.unlink(filePath, (err) => {
        if (err) {
            logger.error(`Failed to delete file: ${filePath}, Error: ${err.message}`);
        } else {
            logger.info(`Deleted old backup: ${filePath}`);
        }
    });
};

// Export the createBackup function for automatic execution in your server
module.exports = createBackup;

// If this script is run directly, execute the backup
if (require.main === module) {
    logger.info('Running backup manually...');
    createBackup();
}

File: storage2/backup.go
// backup.go

package main

import (
	"fmt"
	"github.com/sirupsen/logrus"
	"os"
	"path/filepath"
	"strings"
	"time"
)

// CreateBackup - runs daily backup
func CreateBackup() {
	dbName := os.Getenv("DB_NAME")

	now := time.Now()
	dateStr := now.Format("2006-01-02")
	backupFilename := fmt.Sprintf("user_pokemon_backup_%s.sql", dateStr)

	backupFilePath := filepath.Join("backups", backupFilename)
	_ = os.MkdirAll("backups", os.ModePerm)

	myCnfPath := filepath.Join(".", "my.cnf")
	dumpCmd := fmt.Sprintf("mysqldump --defaults-extra-file=%s %s > %s", myCnfPath, dbName, backupFilePath)

	logrus.Info("Executing backup command: %s\n", dumpCmd)
	if err := runShellCommand(dumpCmd); err != nil {
		logrus.Info("Failed to create backup: %v\n", err)
		return
	}

	logrus.Info("Backup created successfully: %s\n", backupFilePath)
	manageRetention()
}

func manageRetention() {
	dailyRetentionDays := 30
	monthlyRetentionMonths := 12
	yearlyRetentionYears := 5

	backupsDir := "backups"
	now := time.Now()

	files, err := os.ReadDir(backupsDir)
	if err != nil {
		logrus.Info("Failed to read backups dir: %v\n", err)
		return
	}

	for _, f := range files {
		if f.IsDir() {
			continue
		}
		name := f.Name()
		if !strings.HasSuffix(name, ".sql") {
			continue
		}

		parts := strings.Split(strings.TrimSuffix(name, ".sql"), "_")
		if len(parts) < 4 {
			logrus.Info("Filename does not match expected format: %s\n", name)
			continue
		}

		dateStr := parts[len(parts)-1]
		fileDate, err := time.Parse("2006-01-02", dateStr)
		if err != nil {
			logrus.Info("Error parsing date from file %s: %v\n", name, err)
			continue
		}

		isYearly := (fileDate.Month() == time.January && fileDate.Day() == 1)
		isMonthly := (fileDate.Day() == 1 && !isYearly)
		isDaily := (!isYearly && !isMonthly)

		ageDays := int(now.Sub(fileDate).Hours() / 24)
		ageMonths := (int(now.Year())-int(fileDate.Year()))*12 + int(now.Month()) - int(fileDate.Month())
		ageYears := now.Year() - fileDate.Year()

		deleteFile := false
		if isDaily && ageDays > dailyRetentionDays {
			deleteFile = true
		} else if isMonthly && ageMonths > monthlyRetentionMonths {
			deleteFile = true
		} else if isYearly && ageYears > yearlyRetentionYears {
			deleteFile = true
		}

		if deleteFile {
			fullPath := filepath.Join(backupsDir, name)
			if err := os.Remove(fullPath); err != nil {
				logrus.Info("Failed to delete old backup %s: %v\n", name, err)
			} else {
				logrus.Info("Deleted old backup: %s\n", name)
			}
		}
	}
	logrus.Println("Finished managing backup retention.")
}

func runShellCommand(cmd string) error {
	// Example shell call (commented out):
	// out, err := exec.Command("sh", "-c", cmd).CombinedOutput()
	// logrus.Println(string(out))
	// return err
	return nil
}
